internal/xds/rbac/rbac_engine_test.go

```diff
@@ -380,226 +529,374 @@ func (s) TestRBACEngine(t *testing.T) {
+						},
+						destinationPort: 8080,
+						peerInfo: &peer.Peer{
+							Addr: &addr{ipAddress: "10.0.0.0"},
+						},
+					},
+					wantStatusCode: codes.PermissionDenied,
 				},
 			},
 		},
 	}
+
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			// Instantiate the rbacEngine with different configurations that
+			// Instantiate the chainedRBACEngine with different configurations that are
 			// interesting to test and to query.
-			rbacEngine, err := NewEngine(test.rbacConfig)
+			cre, err := NewChainEngine(test.rbacConfigs)
 			if err != nil {
 				t.Fatalf("Error constructing RBAC Engine: %v", err)
 			}
-			// Query that created RBAC Engine with different args to see if the
-			// RBAC Engine configured a certain way works as intended.
-			for _, queryToRBACEngine := range test.rbacQueries {
-				// The matchingPolicyName returned will be empty in the case of
-				// no matching policy. Thus, matchingPolicyName can also be used
-				// to test the "error" condition of no matching policies.
-				matchingPolicyName, matchingPolicyFound := rbacEngine.FindMatchingPolicy(queryToRBACEngine.rpcData)
-				if matchingPolicyFound != queryToRBACEngine.wantMatch || matchingPolicyName != queryToRBACEngine.wantMatchingPolicyName {
-					t.Errorf("FindMatchingPolicy(%+v) returned (%v, %v), want (%v, %v)", queryToRBACEngine.rpcData, matchingPolicyName, matchingPolicyFound, queryToRBACEngine.wantMatchingPolicyName, queryToRBACEngine.wantMatch)
+			// Query the created chain of RBAC Engines with different args to see
+			// if the chain of RBAC Engines configured as such works as intended.
+			for _, data := range test.rbacQueries {
+				// Construct the context with three data points that have enough
+				// information to represent incoming RPC's. This will be how a
+				// user uses this API. A user will have to put MD, PeerInfo, and
+				// the connection the RPC is sent on in the context.
+				ctx := metadata.NewIncomingContext(context.Background(), data.rpcData.md)
+
+				// Make a TCP connection with a certain destination port. The
+				// address/port of this connection will be used to populate the
+				// destination ip/port in RPCData struct. This represents what
+				// the user of ChainEngine will have to place into
+				// context, as this is only way to get destination ip and port.
+				lis, err := net.Listen("tcp", "localhost:"+fmt.Sprint(data.rpcData.destinationPort))
+				if err != nil {
+					t.Fatalf("Error listening: %v", err)
 				}
+				connCh := make(chan net.Conn, 1)
+				go func() {
+					conn, err := lis.Accept()
+					if err != nil {
+						t.Errorf("Error accepting connection: %v", err)
+						return
+					}
+					connCh <- conn
+				}()
+				net.Dial("tcp", lis.Addr().String())
+				conn := <-connCh
+				ctx = SetConnection(ctx, conn)
+				ctx = peer.NewContext(ctx, data.rpcData.peerInfo)
+
+				err = cre.IsAuthorized(ctx, data.rpcData.fullMethod)
+				status, _ := status.FromError(err)
```

Comment:
Since we expect an error compatible with the status package, you could directly call `status.Code` to get the code and compare it with the code that you expect.

```golang
err := cre.IsAuthorized(ctx, data.rpcData.fullMethod)
if gotCode := status.Code(err); gotCode != data.wantStatusCode {
  t.Fatalf(...)
}
```

Reply:
`status` is a package not a struct.  This should work fine as long as you don't name a local variable `status`.

Reply:
Ah, :). Switched.